refactored and moved all interfaces to abc.py

Khwahish29 opened the pull request:
## What was wrong?
Interfaces were scattered across different modules, leading to circular import issues when using them as type hints.
Issue # #228 

## How was it fixed?
Moved all interfaces to a single module, libp2p/abc.py, following the approach used in [py-evm](https://github.com/ethereum/py-evm/blob/3b0bcc21954134e0293e40b1850e6bf16e5f4c27/eth/abc.py). This centralizes interface definitions, making imports cleaner and avoiding circular dependencies.


### To-Do

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/libp2p/py-libp2p/blob/main/newsfragments/README.md)

![put a cute animal picture link inside the parentheses](<https://render.fineartamerica.com/images/rendered/default/poster/8/5.5/break/images-medium-5/brothers-for-life-jeffrey-c-sink.jpg>)

vking45 commented:
hey @pacrob 
@Khwahish29 has done the major heavy lifting of moving almost all the interfaces to `libp2p.abc` file but we have run into one interface which is causing trouble - [`transport_interface`](https://github.com/Khwahish29/py-libp2p/blob/move-to-abc/libp2p/transport/transport_interface.py), since it imports THandler from `transport.typing` which in turns imports from `libp2p.abc`, it is fine to keep the `transport_interface` in it's place or you can guide us on how to move forward from here

mystical-prog commented:
hey @pacrob 
@Khwahish29 has done the major heavy lifting of moving almost all the interfaces to `libp2p.abc` file but we have run into one interface which is causing trouble - [`transport_interface`](https://github.com/Khwahish29/py-libp2p/blob/move-to-abc/libp2p/transport/transport_interface.py), since it imports THandler from `transport.typing` which in turns imports from `libp2p.abc`, it is fine to keep the `transport_interface` in it's place or you can guide us on how to move forward from here

PS : I will fix the docs CI which is failing right now before marking the PR ready for review

dhuseby commented:
Wow 🚀 

Nice job!!

pacrob commented:
I think you should be move the contents of `typing.py` in to the top level `custom_types.py` file. Then we can avoid typing related circular import issues too. Look at doing the same for `libp2p/pubsub/typing.py` too. You can leave the `typing.py` within `tools/async_service` as is - that's a completely separate tool that shouldn't be integrated into libp2p.

Otherwise PR looks great! There are a couple places where the spacing around the section separators could be more consistent (I'll leave a specific comment to show what I mean). 

Thanks for working on this!

pacrob commented on libp2p/abc.py:
```diff
@@ -0,0 +1,867 @@
+        """Close both ends of the stream."""
+
+
+# net_connection_interface
```
Looking for consistent spacing on these comment lines - 2 spaces above and below, like `#net_stream_interface` above.

mystical-prog commented:
thank you so much for the clarification and guidance, we will do the needful and close this issue by today itself and rest assured we will maintain a proper spacing and make it as visually pleasing as possible

seetadev commented:
@mystical-prog  : Thanks Vraj. Kindly share the pull request soon. Will discuss with @pacrob and merge the changes. 

Great  work by @Khwahish29  :)  

> thank you so much for the clarification and guidance, we will do the needful and close this issue by today itself and rest assured we will maintain a proper spacing and make it as visually pleasing as possible

Khwahish29 commented:
Hello! 
We have moved all the interfaces and contents of `typing.py` in to the top level `custom_types.py` file as suggested by @pacrob and adjusted the spacing around the section. We wanted to check how to fix the docs CI issue, as we are unsure whether we need to create a new `.rst` file for `abc.py` by removing all the interface references. 

Any guidance on this would be greatly appreciated!

Thanks :)

mystical-prog commented:
hey @pacrob 
we have fixed the documentation issue and moved all the interfaces to `abc.py` and all types to `custom_types.py`, let us know if there are any changes or fixes to be made :clinking_glasses:

pacrob reviewed (approved):
looks great, thanks!
